frontend/src/metabase/lib/time.js

```diff
@@ -40,9 +43,39 @@ const NUMERIC_UNIT_FORMATS = {
       .startOf("quarter"),
 };
 
+/// Update Moment locale information such as the first day of the week based on their values in the Metabase instance
+/// settings.
+function updateLocaleInfo() {
+  const startOfWeekDayName = MetabaseSettings.get("start-of-week");
+  if (!startOfWeekDayName) {
+    return;
+  }
+
+  const START_OF_WEEK_DAYS = [
+    "sunday",
+    "monday",
+    "tuesday",
+    "wednesday",
+    "thursday",
+    "friday",
+    "saturday",
+  ];
+  const startOfWeekDayNumber =
+    _.indexOf(START_OF_WEEK_DAYS, startOfWeekDayName) || 0;
+
+  moment.updateLocale(moment.locale(), {
+    week: {
+      // Moment.js dow range Sunday (0) - Saturday (6)
+      dow: startOfWeekDayNumber,
+    },
+  });
+}
+
 // only attempt to parse the timezone if we're sure we have one (either Z or ±hh:mm or +-hhmm)
 // moment normally interprets the DD in YYYY-MM-DD as an offset :-/
 export function parseTimestamp(value, unit = null, local = false) {
+  updateLocaleInfo();
```

Comment:
It feels like this should happen after the settings are retrieved from the server rather than before parsing a timestamp. Though, I'd bet updating locale is fast enough so that either way is fine.